<a href="https://colab.research.google.com/github/DeHartJA/Invoices-analysis/blob/main/Python_Module_6_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
df = pd.read_csv('/content/drive/MyDrive/invoices.csv')

<ipython-input-2-03087b5b5325>:2: DtypeWarning: Columns (12) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('/content/drive/MyDrive/invoices.csv')


The purpose of this Analysis is to do apply these things to the data base


    Fix standardization errors, where applicable
    Fix business logic errors, where applicable
    Impute null values, where possible
    Drop columns with many null values, where applicable
    Drop rows with null values, where applicable
    Drop duplicate rows, where applicable
    Identify outliers and fix them where applicable


The Reason we want to do the above mentioned things is so we can answer the following questions,

    Which customer has the most invoices?
    Which item has the highest total quantity sold?
    Which item has generated the most total profit?
    Which item do customers purchase the most quantity of, on average?



Overview

the first five rows are displayed below, the invoices.csv file has the following columns: InvoiceLineID,InvoiceID,StockItemID,Description,PackageTypeID,Quantity,UnitPrice,TaxRate,TaxAmount, LineProfit,ExtendedPrice,LastEditedBy,LastEditedWhen and Customer.

the coding df.head displays the first five rows of the data.

In [ ]:
df.head()

InvoiceLineID  InvoiceID  StockItemID  \
0              1          1         67.0   
1              2          2         50.0   
2              3          2         10.0   
3              4          3        114.0   
4              5          4        206.0   

                                         Description  PackageTypeID  Quantity  \
0           Ride on toy sedan car (Black) 1/12 scale              7        10   
1  Developer joke mug - old C developers never di...              7         9   
2               USB food flash drive - chocolate bar              7         9   
3                 Superhero action jacket (Blue) XXL              7         3   
4         Permanent marker black 5mm nib (Black) 5mm              7        96   

   UnitPrice  TaxRate  TaxAmount  LineProfit  ExtendedPrice  LastEditedBy  \
0      230.0       15      345.0       850.0         2645.0             7   
1       13.0       15        NaN        76.5         -999.0             7   
2       32.0       15       43.2       180.0          331.2             7   
3       30.0       15        NaN        24.0         -999.0             7   
4        2.7       15        NaN        96.0            NaN             7   

   LastEditedWhen          Customer  
0             NaN     Stuff by Stew  
1  1/1/2013 12:00   Wholesaler Plus  
2             NaN   Wholesaler Plus  
3             NaN   Big Buys Retail  
4  1/1/2013 12:00  Terry's Trinkets

Next, general data about World Wide Importers data,there are 230,548 rows and 14 columns. most of the columns don't have null values however the ones that do like TaxAmount and ExtendedPrice will be adressed later in the analysis.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230548 entries, 0 to 230547
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   InvoiceLineID   230548 non-null  int64  
 1   InvoiceID       230548 non-null  int64  
 2   StockItemID     230087 non-null  float64
 3   Description     229876 non-null  object 
 4   PackageTypeID   230548 non-null  int64  
 5   Quantity        230548 non-null  int64  
 6   UnitPrice       230548 non-null  float64
 7   TaxRate         230548 non-null  int64  
 8   TaxAmount       109623 non-null  float64
 9   LineProfit      230548 non-null  float64
 10  ExtendedPrice   113081 non-null  float64
 11  LastEditedBy    230548 non-null  int64  
 12  LastEditedWhen  6 non-null       object 
 13  Customer        230548 non-null  object 
dtypes: float64(5), int64(6), object(3)
memory usage: 24.6+ MB


Finally, Basic descriptive statistics about the dataframe are shown below, most of the data shows the numbers to be positive numbers which is how it should be however their is the LineProfit column that could in theory show a negative number if they took a loss on an Item.
The TaxRate also has a negative number in its Minimum amount, this can be fixed by multipling the taxrate column by -1 as seen in the code below

In [ ]:
# locates any values in the taxrate and extendedprice columns that are less then o, if its a negative number change it to a positive number
df.loc[df['TaxRate'] < 0, 'TaxRate'] *= -1
df.loc[df['ExtendedPrice'] < 0, 'ExtendedPrice'] *= -1

In [ ]:
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230548.000000  230548.000000  230087.000000  230548.000000   
mean   114127.066880   35177.386891     110.181827       7.073638   
std     65900.142929   20338.920057      63.722514       0.644528   
min         1.000000       1.000000       1.000000       1.000000   
25%     57062.750000   17570.000000      54.000000       7.000000   
50%    114130.500000   35151.000000     111.000000       7.000000   
75%    171203.250000   52766.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230548.000000  230548.000000  230548.000000  109623.000000   
mean       39.252524      45.539361      14.463409     113.328006   
std        55.193934     139.510031       3.904056     221.866696   
min         1.000000       0.660000     -15.000000       0.380000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

         LineProfit  ExtendedPrice   LastEditedBy  
count  230548.00000  113081.000000  230548.000000  
mean      375.38894     740.092626      10.800480  
std       753.17041    1705.229089       5.509623  
min      -645.00000    -999.000000       2.000000  
25%        51.00000      89.700000       6.000000  
50%       120.00000     248.400000      11.000000  
75%       390.00000     919.080000      16.000000  
max      9200.00000  114954.000000      20.000000

Part of the cleaning process for this data set is analysing the customer column and making sure everything is spelt the same way, in the code below you can see that Stuff by Stew has 3 different spellings.

value_counts is used to see all the options of a column to check for mistakes.

In [ ]:
df['Customer'].value_counts()

Terry's Trinkets       85593
Stuff by Stew          62752
Stuf by Stew           31984
Wholesaler Plus        15735
Big Buys Retail        13461
Terry's Trinket         9483
stuff by stew           7068
Dennis Distribution     4472
Name: Customer, dtype: int64

To fix the mistakes and make everything match, we can create a dictionary and set the value to the correct value that we want displayed.

Question 5

In [ ]:
# this code replaces the first string ei stuf by stew with a new value , Stuff by Stew to standerize the values
df['Customer'] = df['Customer'].replace({"Stuf by Stew":"Stuff by Stew",
                                         "stuff by stew":"Stuff by Stew",
                                         "Terry's Trinket":"Terry's Trinkets",
                                         })

After doing that we do another value_counts to make sure those changes are implemented

In [ ]:
df['Customer'].value_counts()

Stuff by Stew          101804
Terry's Trinkets        95076
Wholesaler Plus         15735
Big Buys Retail         13461
Dennis Distribution      4472
Name: Customer, dtype: int64

In [ ]:
df['Customer'].unique()

array(['Stuff by Stew', 'Wholesaler Plus', 'Big Buys Retail',
       "Terry's Trinkets", 'Dennis Distribution'], dtype=object)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230548 entries, 0 to 230547
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   InvoiceLineID   230548 non-null  int64  
 1   InvoiceID       230548 non-null  int64  
 2   StockItemID     230087 non-null  float64
 3   Description     229876 non-null  object 
 4   PackageTypeID   230548 non-null  int64  
 5   Quantity        230548 non-null  int64  
 6   UnitPrice       230548 non-null  float64
 7   TaxRate         230548 non-null  int64  
 8   TaxAmount       109623 non-null  float64
 9   LineProfit      230548 non-null  float64
 10  ExtendedPrice   113081 non-null  float64
 11  LastEditedBy    230548 non-null  int64  
 12  LastEditedWhen  6 non-null       object 
 13  Customer        230548 non-null  object 
dtypes: float64(5), int64(6), object(3)
memory usage: 24.6+ MB


In order to answer the question of which customer has the most invoices , we first need to group by customer and invoiceID, by using the nunique function we can check for each unique invoice from that customer,next we find the max number of those invoices and which customer had the most unique invoices. in this case the answer is Stuff by Stew with 31,198 invoices.

In [ ]:

# Count the number of unique invoices for each customer
unique_invoices_per_customer = df.groupby('Customer')['InvoiceID'].nunique()

# Determine the customer with the most unique invoices
most_unique_invoices_customer = unique_invoices_per_customer.idxmax()
num_unique_invoices = unique_invoices_per_customer.max()

print("Customer with the most unique invoices:", most_unique_invoices_customer)
print("Number of unique invoices:", num_unique_invoices)


Customer with the most unique invoices: Stuff by Stew
Number of unique invoices: 31198


Sometimes the information from the data is incomplete and there is not a good way to impute that data, in the case of LastEditedWhen Column, we can't use other data to fill that column so our best course of action is to drop the column all together,

The code shown below does just that.

Question 6

In [ ]:
df.drop(columns = ['LastEditedWhen'], inplace = True)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230548 entries, 0 to 230547
Data columns (total 13 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   InvoiceLineID  230548 non-null  int64  
 1   InvoiceID      230548 non-null  int64  
 2   StockItemID    230087 non-null  float64
 3   Description    229876 non-null  object 
 4   PackageTypeID  230548 non-null  int64  
 5   Quantity       230548 non-null  int64  
 6   UnitPrice      230548 non-null  float64
 7   TaxRate        230548 non-null  int64  
 8   TaxAmount      109623 non-null  float64
 9   LineProfit     230548 non-null  float64
 10  ExtendedPrice  113081 non-null  float64
 11  LastEditedBy   230548 non-null  int64  
 12  Customer       230548 non-null  object 
dtypes: float64(5), int64(6), object(2)
memory usage: 22.9+ MB


When we want to fill in missing data we first have to identify ways that can be done, in this case we can use the StockItemID and Description, if we know what both of these are then we can fill in description if we know what the stockitemid is and vice versa

Question 8

In [ ]:
# drop rows in the StockItemID and Description columns if either column have null values.
df.dropna(subset=['StockItemID', 'Description'])

InvoiceLineID  InvoiceID  StockItemID  \
0                   1          1         67.0   
1                   2          2         50.0   
2                   3          2         10.0   
3                   4          3        114.0   
4                   5          4        206.0   
...               ...        ...          ...   
230543          36558      11285        139.0   
230544          21110       6502          9.0   
230545          34087      10504        179.0   
230546         103435      31836         20.0   
230547         192963      59503         70.0   

                                              Description  PackageTypeID  \
0                Ride on toy sedan car (Black) 1/12 scale              7   
1       Developer joke mug - old C developers never di...              7   
2                    USB food flash drive - chocolate bar              7   
3                      Superhero action jacket (Blue) XXL              7   
4              Permanent marker black 5mm nib (Black) 5mm              7   
...                                                   ...            ...   
230543                        Furry animal socks (Pink) M             10   
230544                      USB food flash drive - banana              7   
230545              Shipping carton (Brown) 229x229x229mm              7   
230546       DBA joke mug - you might be a DBA if (White)              7   
230547           Ride on toy sedan car (Green) 1/12 scale              7   

        Quantity  UnitPrice  TaxRate  TaxAmount  LineProfit  ExtendedPrice  \
0             10     230.00       15     345.00       850.0        2645.00   
1              9      13.00       15        NaN        76.5         999.00   
2              9      32.00       15      43.20       180.0         331.20   
3              3      30.00       15        NaN        24.0         999.00   
4             96       2.70       15        NaN        96.0            NaN   
...          ...        ...      ...        ...         ...            ...   
230543        36       5.00       15      27.00       126.0         207.00   
230544         2      32.00       15        NaN        40.0            NaN   
230545       125       1.05       15        NaN        62.5            NaN   
230546         1      13.00       15       1.95         8.5          14.95   
230547         5     230.00       15        NaN       425.0            NaN   

        LastEditedBy          Customer  
0                  7     Stuff by Stew  
1                  7   Wholesaler Plus  
2                  7   Wholesaler Plus  
3                  7   Big Buys Retail  
4                  7  Terry's Trinkets  
...              ...               ...  
230543            20   Wholesaler Plus  
230544            10     Stuff by Stew  
230545            16  Terry's Trinkets  
230546            20   Big Buys Retail  
230547             4     Stuff by Stew  

[229876 rows x 13 columns]

In [ ]:
# find the value counts of the StockItemID column
df['StockItemID'].value_counts()

120.0    1134
104.0    1131
167.0    1128
13.0     1127
88.0     1119
         ... 
221.0     131
225.0     126
224.0     119
227.0     118
223.0     117
Name: StockItemID, Length: 227, dtype: int64

The code below matches up StockItemID and Description to Impute the null values

In [ ]:
# define the imput description function, if StockItemID is not null and description is null, fill in the description

# Function to impute Description based on StockItemID
def impute_description(row):
    if pd.notnull(row['StockItemID']):
        description = df.loc[df['StockItemID'] == row['StockItemID'], 'Description'].values
        if len(description) > 0:
            return description[0]
    return None

# Apply the function to impute Description
df['Description'] = df.apply(impute_description, axis=1)

print(df)


        InvoiceLineID  InvoiceID  StockItemID  \
0                   1          1         67.0   
1                   2          2         50.0   
2                   3          2         10.0   
3                   4          3        114.0   
4                   5          4        206.0   
...               ...        ...          ...   
230543          36558      11285        139.0   
230544          21110       6502          9.0   
230545          34087      10504        179.0   
230546         103435      31836         20.0   
230547         192963      59503         70.0   

                                              Description  PackageTypeID  \
0                Ride on toy sedan car (Black) 1/12 scale              7   
1       Developer joke mug - old C developers never di...              7   
2                    USB food flash drive - chocolate bar              7   
3                      Superhero action jacket (Blue) XXL              7   
4              Permanent marker

After applying the above code we can drop the rows for StockItemID and Description if both are Null

In [ ]:
# drop rows that show null values in both the StockItemID and the Description column
df.dropna(subset=['StockItemID', 'Description'])

InvoiceLineID  InvoiceID  StockItemID  \
0                   1          1         67.0   
1                   2          2         50.0   
2                   3          2         10.0   
3                   4          3        114.0   
4                   5          4        206.0   
...               ...        ...          ...   
230543          36558      11285        139.0   
230544          21110       6502          9.0   
230545          34087      10504        179.0   
230546         103435      31836         20.0   
230547         192963      59503         70.0   

                                              Description  PackageTypeID  \
0                Ride on toy sedan car (Black) 1/12 scale              7   
1       Developer joke mug - old C developers never di...              7   
2                    USB food flash drive - chocolate bar              7   
3                      Superhero action jacket (Blue) XXL              7   
4              Permanent marker black 5mm nib (Black) 5mm              7   
...                                                   ...            ...   
230543                        Furry animal socks (Pink) M             10   
230544                      USB food flash drive - banana              7   
230545              Shipping carton (Brown) 229x229x229mm              7   
230546       DBA joke mug - you might be a DBA if (White)              7   
230547           Ride on toy sedan car (Green) 1/12 scale              7   

        Quantity  UnitPrice  TaxRate  TaxAmount  LineProfit  ExtendedPrice  \
0             10     230.00       15     345.00       850.0        2645.00   
1              9      13.00       15        NaN        76.5         999.00   
2              9      32.00       15      43.20       180.0         331.20   
3              3      30.00       15        NaN        24.0         999.00   
4             96       2.70       15        NaN        96.0            NaN   
...          ...        ...      ...        ...         ...            ...   
230543        36       5.00       15      27.00       126.0         207.00   
230544         2      32.00       15        NaN        40.0            NaN   
230545       125       1.05       15        NaN        62.5            NaN   
230546         1      13.00       15       1.95         8.5          14.95   
230547         5     230.00       15        NaN       425.0            NaN   

        LastEditedBy          Customer  
0                  7     Stuff by Stew  
1                  7   Wholesaler Plus  
2                  7   Wholesaler Plus  
3                  7   Big Buys Retail  
4                  7  Terry's Trinkets  
...              ...               ...  
230543            20   Wholesaler Plus  
230544            10     Stuff by Stew  
230545            16  Terry's Trinkets  
230546            20   Big Buys Retail  
230547             4     Stuff by Stew  

[230087 rows x 13 columns]

Question 11

In [ ]:
df.describe()

InvoiceLineID      InvoiceID    StockItemID  PackageTypeID  \
count  230548.000000  230548.000000  230087.000000  230548.000000   
mean   114127.066880   35177.386891     110.181827       7.073638   
std     65900.142929   20338.920057      63.722514       0.644528   
min         1.000000       1.000000       1.000000       1.000000   
25%     57062.750000   17570.000000      54.000000       7.000000   
50%    114130.500000   35151.000000     111.000000       7.000000   
75%    171203.250000   52766.000000     165.000000       7.000000   
max    228265.000000   70510.000000     227.000000      10.000000   

            Quantity      UnitPrice        TaxRate      TaxAmount  \
count  230548.000000  230548.000000  230548.000000  109623.000000   
mean       39.252524      45.539361      14.977272     113.328006   
std        55.193934     139.510031       0.336342     221.866696   
min         1.000000       0.660000      10.000000       0.380000   
25%         5.000000      13.000000      15.000000      14.400000   
50%        10.000000      18.000000      15.000000      34.500000   
75%        60.000000      32.000000      15.000000     129.600000   
max      3150.000000    1899.000000      15.000000   14994.000000   

         LineProfit  ExtendedPrice   LastEditedBy  
count  230548.00000  113081.000000  230548.000000  
mean      375.38894     878.050231      10.800480  
std       753.17041    1638.465489       5.509623  
min      -645.00000       2.880000       2.000000  
25%        51.00000     119.600000       6.000000  
50%       120.00000     296.700000      11.000000  
75%       390.00000     999.000000      16.000000  
max      9200.00000  114954.000000      20.000000

In [ ]:
df['TaxRate'].mean()

14.97727154432049

This code multiplies the unit price by quantity and taxrate and then divids it by 100 to get the TaxAmount, this code also fills in any null values in the TaxAmount column so there are no null values in that column now.

In [ ]:
df['TaxAmount'] = df['TaxAmount'].fillna((df['UnitPrice'] * df['Quantity'] * df['TaxRate'] / 100))

In [ ]:
df['TaxAmount'].mean()

112.97782939995145

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230548 entries, 0 to 230547
Data columns (total 13 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   InvoiceLineID  230548 non-null  int64  
 1   InvoiceID      230548 non-null  int64  
 2   StockItemID    230087 non-null  float64
 3   Description    230087 non-null  object 
 4   PackageTypeID  230548 non-null  int64  
 5   Quantity       230548 non-null  int64  
 6   UnitPrice      230548 non-null  float64
 7   TaxRate        230548 non-null  int64  
 8   TaxAmount      230548 non-null  float64
 9   LineProfit     230548 non-null  float64
 10  ExtendedPrice  113081 non-null  float64
 11  LastEditedBy   230548 non-null  int64  
 12  Customer       230548 non-null  object 
dtypes: float64(5), int64(6), object(2)
memory usage: 22.9+ MB


This code multiplies the unitprice by quantity and then adds the taxamount to create the extendedprice, also fill in any null values

In [ ]:
df['ExtendedPrice'] = df['ExtendedPrice'].fillna((df['UnitPrice'] * df['Quantity']) + df['TaxAmount'])

In [ ]:
df['ExtendedPrice'].mean()

871.5749323134445

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230548 entries, 0 to 230547
Data columns (total 13 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   InvoiceLineID  230548 non-null  int64  
 1   InvoiceID      230548 non-null  int64  
 2   StockItemID    230087 non-null  float64
 3   Description    230087 non-null  object 
 4   PackageTypeID  230548 non-null  int64  
 5   Quantity       230548 non-null  int64  
 6   UnitPrice      230548 non-null  float64
 7   TaxRate        230548 non-null  int64  
 8   TaxAmount      230548 non-null  float64
 9   LineProfit     230548 non-null  float64
 10  ExtendedPrice  230548 non-null  float64
 11  LastEditedBy   230548 non-null  int64  
 12  Customer       230548 non-null  object 
dtypes: float64(5), int64(6), object(2)
memory usage: 22.9+ MB


In [ ]:
# Get the average and standard deviations of Fare
average_qty = df['Quantity'].mean()
std_dev_qty = df['Quantity'].std()



In [ ]:
df['Z-Score'] = (df['Quantity'] - average_qty) / std_dev_qty
df.head()

InvoiceLineID  InvoiceID  StockItemID  \
0              1          1         67.0   
1              2          2         50.0   
2              3          2         10.0   
3              4          3        114.0   
4              5          4        206.0   

                                         Description  PackageTypeID  Quantity  \
0           Ride on toy sedan car (Black) 1/12 scale              7        10   
1  Developer joke mug - old C developers never di...              7         9   
2               USB food flash drive - chocolate bar              7         9   
3                 Superhero action jacket (Blue) XXL              7         3   
4         Permanent marker black 5mm nib (Black) 5mm              7        96   

   UnitPrice  TaxRate  TaxAmount  LineProfit  ExtendedPrice  LastEditedBy  \
0      230.0       15     345.00       850.0        2645.00             7   
1       13.0       15      17.55        76.5         999.00             7   
2       32.0       15      43.20       180.0         331.20             7   
3       30.0       15      13.50        24.0         999.00             7   
4        2.7       15      38.88        96.0         298.08             7   

           Customer   Z-Score  
0     Stuff by Stew -0.529995  
1   Wholesaler Plus -0.548113  
2   Wholesaler Plus -0.548113  
3   Big Buys Retail -0.656821  
4  Terry's Trinkets  1.028147

In [ ]:
df.loc[df['Z-Score'] > 3]

InvoiceLineID  InvoiceID  StockItemID  \
122               123         59        191.0   
197               198         95        191.0   
298               299        129        196.0   
305               306        132        186.0   
307               308        134        189.0   
...               ...        ...          ...   
230285         171205      52767        183.0   
230377          74832      23009        182.0   
230458         151681      46729        195.0   
230475         217008      67029        201.0   
230510         154951      47750        201.0   

                                              Description  PackageTypeID  \
122       Black and orange fragile despatch tape 48mmx75m              7   
197       Black and orange fragile despatch tape 48mmx75m              7   
298     Black and orange handle with care despatch tap...              7   
305                 Shipping carton (Brown) 457x457x457mm              7   
307                         Clear packaging tape 48mmx75m              7   
...                                                   ...            ...   
230285              Shipping carton (Brown) 480x270x320mm              7   
230377              Shipping carton (Brown) 457x279x279mm              7   
230458  Black and orange handle with care despatch tap...              7   
230475        Red and white urgent despatch tape 48mmx75m              7   
230510        Red and white urgent despatch tape 48mmx75m              7   

        Quantity  UnitPrice  TaxRate  TaxAmount  LineProfit  ExtendedPrice  \
122          288       3.70       15     159.84       547.2        1225.44   
197          252       3.70       15     139.86       478.8        1072.26   
298          216       4.10       15     132.84       453.6        1018.44   
305          250       2.10       15      78.75       250.0         603.75   
307          234       2.90       15     101.79       280.8         780.39   
...          ...        ...      ...        ...         ...            ...   
230285       250       2.74       15     102.75       385.0         787.75   
230377       250       1.28       15      48.00       170.0         368.00   
230458       216       3.70       15     119.88       518.4         919.08   
230475       240       3.70       15     133.20       504.0        1021.20   
230510       240       3.70       15     133.20       504.0        1021.20   

        LastEditedBy          Customer   Z-Score  
122                9  Terry's Trinkets  4.506790  
197                9     Stuff by Stew  3.854545  
298                2     Stuff by Stew  3.202299  
305                2   Big Buys Retail  3.818309  
307                2  Terry's Trinkets  3.528422  
...              ...               ...       ...  
230285            17  Terry's Trinkets  3.818309  
230377            10     Stuff by Stew  3.818309  
230458             5  Terry's Trinkets  3.202299  
230475            18     Stuff by Stew  3.637129  
230510            17   Big Buys Retail  3.637129  

[6002 rows x 14 columns]

This code creates a new column for zscore for the quantity and extended price. zscore is use to identify outliers in the data. Anything with a zscore of greater then 3 is considered an outlier.

In [ ]:
import pandas as pd
from scipy.stats import zscore

# Calculate z-scores for 'Quantity' and 'ExtendedPrice'
df['Quantity_zscore'] = zscore(df['Quantity'])
df['ExtendedPrice_zscore'] = zscore(df['ExtendedPrice'])

# Identify outliers based on z-score threshold of 3
outliers_quantity = df[(df['Quantity_zscore'] > 3) | (df['Quantity_zscore'] < -3)]
outliers_extendedprice = df[(df['ExtendedPrice_zscore'] > 3) | (df['ExtendedPrice_zscore'] < -3)]

print("Outliers in Quantity column:")
print(outliers_extendedprice)



Outliers in Quantity column:
        InvoiceLineID  InvoiceID  StockItemID  \
96                 97         45        164.0   
111               112         53        170.0   
124               125         59        167.0   
256               257        116        167.0   
289               290        126        158.0   
...               ...        ...          ...   
230065         151133      46554        158.0   
230079         219963      67961        167.0   
230298         217447      67164        158.0   
230407         209037      64540        158.0   
230483           1708        598        215.0   

                                     Description  PackageTypeID  Quantity  \
96            32 mm Double sided bubble wrap 50m              7        50   
111     20 mm Anti static bubble wrap (Blue) 50m              7        90   
124     10 mm Anti static bubble wrap (Blue) 50m              7        60   
256     10 mm Anti static bubble wrap (Blue) 50m              7        70 

In [ ]:
import pandas as pd

# Assuming df is your DataFrame containing the dataset

# Check for duplicates using both InvoiceID and InvoiceLineID columns
duplicate_rows = df[df.duplicated(subset=['InvoiceID', 'InvoiceLineID'], keep=False)]

# Display the duplicate rows
print("Duplicate rows based on InvoiceID and InvoiceLineID:")
print(duplicate_rows)


Duplicate rows based on InvoiceID and InvoiceLineID:
        InvoiceLineID  InvoiceID  StockItemID  \
4                   5          4        206.0   
75                 76         35         60.0   
273               274        122         12.0   
334               335        154         16.0   
411               412        178        117.0   
...               ...        ...          ...   
230543          36558      11285        139.0   
230544          21110       6502          9.0   
230545          34087      10504        179.0   
230546         103435      31836         20.0   
230547         192963      59503         70.0   

                                              Description  PackageTypeID  \
4              Permanent marker black 5mm nib (Black) 5mm              7   
75      RC toy sedan car with remote control (Blue) 1/...              7   
273                          USB food flash drive - donut              7   
334            DBA joke mug - mind if I join you? (Whi

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230548 entries, 0 to 230547
Data columns (total 16 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   InvoiceLineID         230548 non-null  int64  
 1   InvoiceID             230548 non-null  int64  
 2   StockItemID           230087 non-null  float64
 3   Description           230087 non-null  object 
 4   PackageTypeID         230548 non-null  int64  
 5   Quantity              230548 non-null  int64  
 6   UnitPrice             230548 non-null  float64
 7   TaxRate               230548 non-null  int64  
 8   TaxAmount             230548 non-null  float64
 9   LineProfit            230548 non-null  float64
 10  ExtendedPrice         230548 non-null  float64
 11  LastEditedBy          230548 non-null  int64  
 12  Customer              230548 non-null  object 
 13  Z-Score               230548 non-null  float64
 14  Quantity_zscore       230548 non-null  float64
 15  

In [ ]:
# Group by 'Description' and sum up the quantities
quantity_sold = df.groupby('Description')['Quantity'].sum()

# Find the item with the highest total quantity sold
item_highest_quantity = quantity_sold.idxmax()
highest_quantity = quantity_sold.max()

print("Item with the highest total quantity sold:", item_highest_quantity)
print("Highest total quantity sold:", highest_quantity)

Item with the highest total quantity sold: Black and orange fragile despatch tape 48mmx75m
Highest total quantity sold: 209088


In [ ]:
total_profit = df.groupby('Description') ['LineProfit'].sum()

item_highest_profit = total_profit.idxmax()
highest_profit = total_profit.max()

print(" Item with the highest total Profit sold", item_highest_profit)
print("Highest Profit", highest_profit)

 Item with the highest total Profit sold 20 mm Double sided bubble wrap 50m
Highest Profit 5323120.0


In [ ]:
# Group by 'Description' and calculate the average quantity
average_quantity = df.groupby('Description')['Quantity'].mean()

# Find the item with the highest average quantity
item_highest_average_quantity = average_quantity.idxmax()
highest_average_quantity = average_quantity.max()

print("Item purchased in greatest quantities in a single order, on average:", item_highest_average_quantity)
print("Average quantity purchased:", highest_average_quantity)

Item purchased in greatest quantities in a single order, on average: Black and orange fragile despatch tape 48mmx75m
Average quantity purchased: 199.5114503816794


Results:

The answer to which customer has the most invoices is Stuff By Stew, they had over 30,000 unique invoices.
The Item that had the highest total quantity sold was Black and Orange Fragile Despatch Tape 48mmx75mm, with a total of 209088 units sold.
20mm Double sided bubble wrap 50m was the highest total Profit sold.
Finally the Item that customers purchased the most quantity of on average was the Black and orange fragile despatch tape 48mmx75mm, with an average of almost 200 units